### ISSUE TITLE
arrow function. This type is incompatible with EventHandler...

### ISSUE THREAD

**Issue by Gozala** [facebook/flow#1413]
Flow incorrectly fails on following code:

``` js
/* @flow */

const receive = (worker/*:Worker*/, handler/*:Function*/)/*:void*/ => {
  const onMessage = event => {
    event.target.removeEventListener("message", onMessage);
    handler(event.data);
  };

  worker.addEventListener("message",  onMessage);
}
```

```
src/listener.js:9
  9:   worker.addEventListener("message",  onMessage);
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call of method `addEventListener`
  9:   worker.addEventListener("message",  onMessage);
                                           ^^^^^^^^^ arrow function. This type is incompatible with
 66:     addEventListener(type: string, listener: EventListener, useCapture?: boolean): void;
                                                  ^^^^^^^^^^^^^ union: object type | EventHandler. See: /private/tmp/flow/flowlib_3b5ce42a/dom.js:66
```

**Comment by Gozala** [comment]
It turns out the reason flow actually fails is access to `event.data` that is expected on message event:
https://www.w3.org/TR/2011/WD-webmessaging-20110317/#dom-messageevent-data

But the error produces by flow is not only invalid but also completely misleading.

**Comment by Gozala** [comment]
Presumably problem is the interface definitions for EventTarget / EventListener:
https://github.com/facebook/flow/blob/master/lib/dom.js#L61-L70

I guess it would make sense to change EventTarget to:

```js
declare class EventTarget <event:Event> {
    removeEventListener(type: string, listener: EventListener<event>, useCapture?: boolean): void;
    addEventListener(type: string, listener: EventListener<event>, useCapture?: boolean): void;
    detachEvent?: (type: string, listener: EventListener<event>) => void;
    attachEvent?: (type: string, listener: EventListener<event>) => void;
    dispatchEvent(evt: event): boolean;
}
```

And then also make each `EventTarget` also polymorphic, so `Worker`:
https://github.com/facebook/flow/blob/master/lib/bom.js#L236-L242
will also change to something like:

```js
type WorkerEvent <data>
  = MessageEvent<data>
  | ErrorEvent

declare class Worker <data> extends EventTarget <WorkerEvent>  {
    constructor(stringUrl: string): void;
    onerror: (ev: Event) => any;
    onmessage: (ev: Event) => any;
    postMessage(message: any, ports?: any): void;
    terminate(): void;
}
```

**Comment by Gozala** [comment]
Although unless I'm mistaken changes described above would mean that function passed to `addEventListener` would have to handle any kind of event that specific EventTarget may dispatch ignoring `type` completely which is wrong and presumably annoying. So what you really want is to refine `Event` for  `EventListener` by `type` but I don't think that's currently possible.

**Comment by claramunt** [comment]
this should be fixed with #7097

**Comment by panagosg7** [comment]
I have confirmed that after #7097 the following code typechecks:
```js
const receive = (worker/*:Worker*/, handler/*:Function*/)/*:void*/ => {
  const onMessage = (event: MessageEvent) => {
    event.target.removeEventListener("message", onMessage);
    handler(event.data);
  };

  worker.addEventListener("message",  onMessage);
}
```
